src/Avalonia.Base/Utilities/DeferredSetter.cs

```diff
@@ -0,0 +1,135 @@
+            }
+            setRecords[property].Notifying = true;
+
+            return Disposable.Create(() => setRecords[property].Notifying = false);
+        }
+
+        /// <summary>
+        /// Check if the property is currently notifying listeners.
+        /// </summary>
+        /// <param name="property">The property.</param>
+        /// <returns>If the property is currently notifying listeners.</returns>
+        internal bool IsNotifying(TProperty property)
+            => setRecords.TryGetValue(property, out var value) && value.Notifying;
+
+        /// <summary>
+        /// Add a pending assignment for the property.
+        /// </summary>
+        /// <param name="property">The property.</param>
+        /// <param name="value">The value to assign.</param>
+        internal void AddPendingSet(TProperty property, TValue value)
+        {
+            Contract.Requires<InvalidOperationException>(IsNotifying(property));
+            if (!setRecords.ContainsKey(property))
+            {
+                setRecords[property] = new SettingStatus();
+            }
+            setRecords[property].PendingValues.Enqueue(value);
+        }
+
+        /// <summary>
+        /// Checks if there are any pending assignments for the property.
+        /// </summary>
+        /// <param name="property">The property to check.</param>
+        /// <returns>If the property has any pending assignments.</returns>
+        internal bool HasPendingSet(TProperty property)
+        {
+            return setRecords.ContainsKey(property) && setRecords[property].PendingValues.Count != 0;
+        }
+
+        /// <summary>
+        /// Gets the first pending assignment for the property.
+        /// </summary>
+        /// <param name="property">The property to check.</param>
+        /// <returns>The first pending assignment for the property.</returns>
+        internal TValue GetFirstPendingSet(TProperty property)
+        {
+            return setRecords[property].PendingValues.Dequeue();
+        }
+
+        /// <summary>
+        /// Set the property and notify listeners while ensuring we don't get into a stack overflow as happens with #855 and #824
+        /// </summary>
+        /// <param name="property">The property to set.</param>
+        /// <param name="setterCallback">
+        /// A callback that actually sets the property.
+        /// The first parameter is the value to set, and the second is a wrapper that takes a callback that sends the property-changed notification.
+        /// </param>
+        /// <param name="value">The value to try to set.</param>
+        /// <param name="pendingSetCondition">A predicate to filter what possible values should be added as pending sets (i.e. only values not equal to the current value).</param>
+        public void SetAndNotify(TProperty property, Action<TValue, Action<Action>> setterCallback, TValue value, Predicate<TValue> pendingSetCondition)
```

Comment:
Nit: Again, I can't see a good part of these lines on GitHub or in Visual Studio.